modules/kinesis/src/main/scala/com/snowplowanalytics/snowplow/enrich/kinesis/Sink.scala

```diff
@@ -224,6 +222,50 @@ object Sink {
       prre
     }
 
+  /**
+   * The result of trying to write a batch to kinesis
+   *  @param toRetry Records to retry, either because of throttling or an internal error
+   *  @param wasThrottled Whether at least one of retries is because of throttling
+   *  @param exampleInternalError A message to help with logging
+   */
+  private case class TryBatchResult(
+    toRetry: Vector[PutRecordsRequestEntry],
+    wasThrottled: Boolean,
+    exampleInternalError: Option[String]
+  ) {
+    def wasSuccessful: Boolean =
+      toRetry.isEmpty || wasThrottled
+  }
+
+  private object TryBatchResult {
+
+    implicit private def tryBatchResultMonoid: Monoid[TryBatchResult] =
+      new Monoid[TryBatchResult] {
+        override val empty: TryBatchResult = TryBatchResult(Vector.empty, false, None)
+        override def combine(x: TryBatchResult, y: TryBatchResult): TryBatchResult =
+          TryBatchResult(x.toRetry ++ y.toRetry, x.wasThrottled || y.wasThrottled, x.exampleInternalError.orElse(y.exampleInternalError))
+      }
+
+    def build(records: List[PutRecordsRequestEntry], prr: PutRecordsResult): TryBatchResult =
+      if (prr.getFailedRecordCount.toInt =!= 0)
+        records
+          .zip(prr.getRecords.asScala)
+          .map {
+            case (orig, recordResult) =>
+              Option(recordResult.getErrorCode) match {
+                case None =>
+                  Monoid[TryBatchResult].empty
+                case Some("ProvisionedThroughputExceededException") =>
+                  TryBatchResult(Vector(orig), true, None)
+                case Some(_) =>
+                  TryBatchResult(Vector(orig), false, Option(recordResult.getErrorMessage))
+              }
+          }
+          .combineAll
```

Comment:
Neat: we iterate through the list twice, when we could iterate only once with a `fold`.